Improve types

Pl217 opened the pull request:
This started as I have noticed that branded IDs aren't done properly in bunch of places in the codebase, but I also did some other improvements along the way. @manelcecs, there are two TODO comments in which I wanted to point out typing issues I'd like you to improve.

manelcecs committed:
Update how 'reporterRefCode' filter works

Now is not exact match but 'isLike'

manelcecs committed:
🏃 Increased performance when select sorted entities

Reorder methods to reduce the select set

Change 'const' for 'let' to just copy the results rather than iterating over it

Pl217 committed:
✏️ Fix typos

Pl217 committed:
Remove usages of `Object.prototype.valueOf()`

Using this method on branded IDs does harm, because
we lose ID branding and get generic `number` type instead

Pl217 committed:
Use branded ID types from hpc-api-core

Some of ID definitions were `number` because they
predated the existence of hpc-api-core

Pl217 committed:
Remove unnecessary usages of `createBrandedValue()`

Pl217 committed:
Remove unnecessary nullish coalescing

The types of these variables are `OrganizationId | null`
and if type is correct, there is no need for nullish coalescing

Pl217 committed:
Use `Map` keys without a loop

Pl217 committed:
Use branded IDs in `Map` keys

Pl217 committed:
Don't use explicit instance classes when fetching data

Pl217 committed:
Don't use `entries()` iterator with `Map`

This iterator is implicit when using `for...of` loop

Pl217 committed:
Use `FlowId` branded type

Pl217 committed:
Remove unnecessary initialization of array value

The same thing is happening in `getOrCreate()` below

Pl217 committed:
Add Strong types

Pl217 committed:
TODO: Rewrite this with strong typing

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

manelcecs committed:
Use 'intersect' new method from Set

This allow to do intersection more performant

manelcecs committed:
Store promises for other entities filters

This allow to handle all request at the same time

Which helps with the time-performance of the endpoint

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review
